tough/src/http.rs

```diff
@@ -180,81 +188,217 @@ impl RetryState {
+                trace!("retryable error during fetch: {}", err);
+                HttpResult::Retryable(err)
+            }
+        }
+    }
+}
+
+/// Checks the HTTP response code and converts a non-successful response code to an error.
+fn parse_response_code(response: reqwest::blocking::Response) -> HttpResult {
+    match response.error_for_status() {
+        Ok(ok) => {
+            trace!("response is success");
+            // http status code indicates success
+            HttpResult::Ok(ok)
+        }
+        // http status is an error
+        Err(err) => match err.status() {
+            None => {
+                // this shouldn't happen, we received this err from the err_for_status function,
+                // so the error should have a status. we cannot consider this a retryable error.
+                trace!("error is fatal (no status): {}", err);
+                HttpResult::Fatal(err)
+            }
+            Some(status) => parse_status_err(err, status),
+        },
+    }
+}
+
+/// Categorizes the the error type based on its HTTP code.
+fn parse_status_err(err: reqwest::Error, status: reqwest::StatusCode) -> HttpResult {
+    if status.is_server_error() {
+        trace!("error is retryable: {}", err);
+        HttpResult::Retryable(err)
+    } else {
+        match status.as_u16() {
+            // some services (like S3) return a 403 when the file is not found
+            403 | 404 => {
+                trace!("error is file not found: {}", err);
+                HttpResult::FileNotFound(err)
+            }
+            _ => {
+                trace!("error is fatal (status): {}", err);
+                HttpResult::Fatal(err)
+            }
+        }
+    }
+}
+
+/// Builds a GET request. If `next_byte` is greater than zero, adds a byte range header to the request.
+fn build_request(
+    client: &Client,
+    next_byte: usize,
+    url: &Url,
+) -> Result<Request, http_error::HttpError> {
     if next_byte == 0 {
         let request = client
             .request(Method::GET, url.as_str())
             .build()
-            .context(error::HttpRequestBuild { url: url.clone() })?;
+            .context(http_error::RequestBuild)?;
         Ok(request)
     } else {
         let header_value_string = format!("bytes={}-", next_byte);
-        let header_value =
-            HeaderValue::from_str(header_value_string.as_str()).context(error::HttpHeader {
+        let header_value = HeaderValue::from_str(header_value_string.as_str()).context(
+            http_error::InvalidHeader {
                 header_value: &header_value_string,
-            })?;
+            },
+        )?;
         let request = client
             .request(Method::GET, url.as_str())
             .header(header::RANGE, header_value)
             .build()
-            .context(error::HttpRequestBuild { url: url.clone() })?;
+            .context(http_error::RequestBuild)?;
         Ok(request)
     }
 }
+
+mod http_error {
```

Comment:
fixup: remove type parameter for transport - tough http.rs rename to_transport_error to from and remove http_error module

Reply:
P.S. my IDE fails me on this one. It turns these red, I think because the macro expansion happens after the usage site. It compiles just fine, but I abhor red things in my code editor which is why I had it there. I've removed it since it's definitely not needed.

Reply:
Addressed in [59899dd](https://github.com/awslabs/tough/compare/3e858e5b5d629576a0c13f3f7b627cd8c49af9a4..59899ddf735ceb7e4edc883a9b1c6a195506f820),